### ISSUE TITLE
Include examples in Sphinx docs

### ISSUE THREAD

**Issue by cdeil** [gammapy/gammapy#172]
We have a folder with Python examples: https://github.com/gammapy/gammapy/tree/master/examples
A table of contents should be auto-generated as part of `build_sphinx` and be put here: https://gammapy.readthedocs.org/en/latest/install.html

(scikit-learn, astro_ML and a few others have this set up, so that also the console or plot output appears on that page)

An alternative would be to move all examples to [IPython notebooks](http://nbviewer.ipython.org/github/gammapy/gammapy-extra/blob/master/notebooks/Index.ipynb) ... I'm not sure yet what is convenient to write and maintain.

**Comment by cdeil** [comment]
Looks like there finally is a package to integrate notebooks better with Sphinx:
https://github.com/spatialaudio/nbsphinx

I haven't tried it yet, i.e. I don't know how well it works or how well-integrated the output is with the rest of the docs (link to API docs? full-text search?)
Probably the other way around would be ideal: author docs in sphinx (so that there are nice simple text files with meaningful diffs in version control) and auto-generate notebooks for people to try out the examples.

If there's someone interested in notebook tooling, exploring options for this and implementing #350 would be very valuable ...

**Comment by adonath** [comment]
Since we have the nice setup with the notebook tutorials now, I'd prefer to have all the (important) examples in notebooks. The other examples in `gammapy/examples` should be converted to a notebook or added to existing notebooks.

**Comment by cdeil** [comment]
As mentioned in the Gammapy call today, I think in principle we want to keep a few Python scripts, because otherwise the message / impression is that Gammapy can only be used from Jupyter notebooks.

I would suggest to ship at least a few Python scripts to users so that they can try out Gammapy from scripts quickly, before they have the skills to write their own because they haven't learned the API yet.

Earlier we did have a structure where some tutorials were Python scripts, and some notebooks:
https://docs.gammapy.org/0.6/tutorials/index.html

And then they were referenced / exposed from the RST docs like this:
https://docs.gammapy.org/0.6/tutorials/background/index.html

Now we have Python some Python script in the docs folder,  mostly used from `plot` directives, I think one (`docs/scripts/significance.py`) with a `literalinclude` directive:
```
docs/scripts/significance.py
docs/irf/plot_edisp.py
docs/irf/plot_fermi_psf.py
docs/astro/source/plot_snr_radius_evolution.py
docs/astro/source/plot_pulsar_spindown.py
docs/astro/source/plot_pwn_evolution.py
docs/astro/source/plot_snr_brightness_evolution.py
docs/astro/population/plot_spiral_arms.py
docs/astro/population/plot_velocity_distributions.py
docs/astro/population/plot_spiral_arm_models.py
docs/astro/population/plot_radial_distributions.py
docs/stats/plot_fc_gauss.py
docs/stats/plot_fc_poisson.py
```
In the rendered version it looks like this for plot directives:
https://docs.gammapy.org/0.9/astro/population/index.html
or like this for literalinclude directive:
https://docs.gammapy.org/0.9/scripts/index.html#write-your-own-cli

In addition we have this, currently not mentioned from the docs:
```
$ ls -1 examples/
README.rst
example_2_gauss.py
example_exptest.py
example_fc_demonstrate_artefact.py
example_fc_gauss_analytical.py
example_fc_poisson_analytical.py
example_fermi_catalogs.py
example_observation_cta.py
```

So @Bultako @adonath or anyone interested - what should we do?

We do need version coupling of Python scripts, just like we do for notebooks, or it'll be a mess. This is already achieved for the Python scripts in the `docs` folder, but not the ones in the `examples` folder. 

Maybe do exactly the same for `examples` as we do for `notebooks`, i.e. ship them to users via `gammapy download`? That solves the version coupling issue, and IMO is more pleasant user experience than the individual "download" buttons that the plot directive generates.

Would we leave the scripts used for the plot directives as-is, or is it useful to ship those to users also into `GAMMAPY_DATA`?

Note that the Sphinx directives work with `../..` and we can make them work with a `GAMMAPY_DATA` env var if needed, i.e. using or making the examples appear in the Sphinx docs isn't difficult either way. The main question IMO is how to ship Python example scripts to users.

In the docs examples should be mentioned here:
https://docs.gammapy.org/0.9/getting-started.html#python-script

Maybe also list them at the end of this page? (probably we'll just have very few)
https://docs.gammapy.org/0.9/tutorials.html

---

Going through the list of Python scripts and deciding which one to delete or rewrite as a notebook would be a separate step / discussion. Maybe do that later. Just a few quick notes.

This might make for a good example (short script, long-running computation): https://github.com/gammasky/cta-dc/blob/master/data/cta_1dc_make_allsky_images.py

We should also link to https://github.com/open-gamma-ray-astro/joint-crab as an example for a large set of scripts to do a full reproducible analysis.

**Comment by cdeil** [comment]
Another option is to use sphinx-gallery, which is what e.g. `scikit-image` or `scikit-learn` use. Example: https://scikit-learn.org/stable/auto_examples/cluster/plot_kmeans_assumptions.html#sphx-glr-auto-examples-cluster-plot-kmeans-assumptions-py

**Comment by Bultako** [comment]
I would propose to deal with this in the same way we already do with the notebooks/tutorials.

SHIPPING 

* Ship the versioned scripts with a new subcommand `gammapy download scripts`.
* The command `gammapy download tutorials` will ship the datasets and the versioned notebooks and scripts.
* Add a `scripts.yaml` file in the folder `examples` as a lookup table index for the scripts, as we already do with `notebooks.yaml` in the `tutorials` folder.

Because we use this kind of LUT with the `scripts.yaml` file, the *.py scripts may be anywhere i.e. as those in inner `docs` subfolder listed above integrated in the docs with `plot` and `literalinclude` directives.

EXPOSING

It could be nice to display the code that is used in the plot directives. I could investigate on how to show/hide the code by clicking in the images or a link `display code`, similar to what is done here http://scopatz.github.io/hiddencode/ 

For the rest of the scripts we could add a subsection `Scripts` in the `Tutorials` section of the docs https://docs.gammapy.org/0.9/tutorials.html to have them listed somewhere. Of course, we can link to a specific python script from anywhere in the docs or use the `literalinclude` directive.

**Comment by cdeil** [comment]
@Bultako - That sounds good.

Note that https://matplotlib.org/devel/plot_directive.html#options has a `include-source` option, so simple show or hide is possible. We just have to discuss what is useful, often plotting code is long and not particularly interesting, so we might not always want to show it.

I'm on vacation starting today until Jan 7, but maybe we could have a 15 min call then where we just make all the small decisions for each script and which plot code to show, and then put up a PR for discussion with the others.

**Comment by cdeil** [comment]
Should be added here: https://github.com/gammapy/gammapy/blob/master/docs/tutorials.rst#scripts

Probably we should change `tutorials.rst` to `tutorials/index.rst` and then add RST pages for the "scripts-based" tutorials. Or alternatively we heavily annotate the Python scripts with comments and docstrings and then don't need the extra RST page.

**Comment by cdeil** [comment]
`exptest.py` is an OK example for something that one would write as a Python script - a long-running simulation. Let's use that as first example how to integrate it into the HTML docs.

https://github.com/gammapy/gammapy/blob/e64b351d5a03735dbc64a072da14d100dabdbd9b/examples/scripts.yaml#L4-L5

It should appear here:
https://docs.gammapy.org/dev/tutorials.html#scripts
When clicking it, the Python code (maybe output) should appear on the HTML page, just downloading the .py file isn't very nice.

Something like this:
https://scikit-image.org/docs/stable/auto_examples/data/plot_scientific.html#sphx-glr-auto-examples-data-plot-scientific-py
Or this:
https://docs.gammapy.org/0.14/spectrum/reflected.html

@Bultako - Do you have a proposal for a technical solution?

If it's not clear how to do it, then I would suggest to change `tutorials.rst` to `tutorials/index.rst` and to add one RST page per example and to link to that. Kind of like how it's generated for notebooks as well.

**Comment by Bultako** [comment]
I guess we could develop a nre Sphinx directive in the same spirit we did to include pieces of the `analysis/config/docs.yaml` in the docs.

https://github.com/gammapy/gammapy/blob/e64b351d5a03735dbc64a072da14d100dabdbd9b/gammapy/utils/docs.py#L41

https://github.com/gammapy/gammapy/blob/e64b351d5a03735dbc64a072da14d100dabdbd9b/gammapy/utils/docs.py#L104

The `gp-howto-hli` directive includes the content of files in this way.
https://docs.gammapy.org/dev/analysis/index.html#data-reduction-and-datasets

So one RST file for each script using this new Sphinx directive to include the code could be ok.

**Comment by cdeil** [comment]
I've added an `examples/survey_map.py` in ef80dc1ebc5b8160ee120da40af8a2f7c1246f9c

How about this?
- Make a folder `docs/tutorials` and move `docs/tutorials.rst` -> `docs/tutorials/index.rst`
- Adapt the notebooks docs machinery to cope (move `docs/notebooks` to `docs/tutorials/notebooks`? Fix links?)
- Add `docs/tutorials/survey_map.rst` (or some other filename, depends on how we want to structure the descriptions of the example scripts), and simply use a `.. literalinclude:: ../../examples/survey_map.py` (see [Sphinx docs](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#showing-code-examples)) to include the script, and have some RST around explaining what the script does and how to execute it.

**Comment by cdeil** [comment]
Implemented by @Bultako in #2550
